src/i2c.rs

```diff
@@ -835,17 +835,70 @@ where
             sr1.addr().bit_is_clear()
         } {}
 
-        // Clear condition by reading SR2
-        self.i2c.sr2.read();
+        if buffer.len() <= 2 {
 
-        // Receive bytes into buffer
-        for c in buffer {
-            *c = self.recv_byte()?;
-        }
+            // Clear the ack flag and set the pos flag to the nack gets sent
+            // after the next recieved byte
+            self.i2c.cr1.modify(|_, w| w.ack().clear_bit().pos().set_bit());
```

Comment:
According to the reference manual:
```
The POS bit must be used only in 2-byte reception configuration in master mode
```
Here you are setting it even when receiving only one byte.

Also this POS trick is a kind of optimization for receiver that always receive exactly 2 bytes, but it should be possible to receive 2 bytes without using the POS trick, and thus, I think we don't need to have special case for len <= 2.